docs/downloader.rst

```diff
@@ -0,0 +1,169 @@
+
+.. -*- coding: utf-8 -*-
+
+.. _downloader:
+
+Downloader Operation
+====================
+
+Audience: developers working on magic-folder itself
+
+**Note:** some of this document describes features that aren't yet implemented; those are noted with (Not implemented).
+
+This describes the general operation of the remote to local synchronization as outlined in "the Leif Design" (see :ref:`Leif's Proposal: Magic-Folder "single-file" snapshot design`).
+
+We do not describe the synchronization of the list of participants in a magic-folder.
+This list will consist of at least:
+
+- an arbitrary ``name`` for each participant
+- a directory read-capability for each participant ("Personal DMD")
+- in this directory is a flattened representation of every file in  the Magic Folder (similar to the Tahoe 1.14.0 design) with each entry pointing to a Snapshot.
+  That is, there are no sub-directories to recurse into: all files are listed by downloading just the "personal DMD" capability of that participant.
+
+"A Snapshot" is a single version of a single file.
+It is represented by an immutable directory and contains:
+
+- ``content``: (optional) a read-only Capability to the actual content of this Snapshot.
+  Deletion Snapshots contain no such key.
+- ``metadata``: information about the Snapshot, a read-only Capability pointing to a JSON-serialized dict containing:
+  - ``snapshot_version``: int(1) currently
+  - ``name``: the name of this snapshot (a mangled relative path)
+  - ``author``: a dict containing:
+    - ``name``: arbitrary name
+    - ``verify_key``: base64-encoded public key of the author
+  - ``parents``: a list of immutable Capability-strings of any parent Snapshots
+  - (not yet) additionally, in the Tahoe metadata for this metadata-capability is a ``magic_folder`` dict with the following keys:
+    - ``author_signature``: base64-encoded signature which signs the content-Capability, metadata-Capability and name
+
+If there are zero parents this is the first version of a file.
+If there is one parent it is a modification.
+If there are two or more parents this version is the resolution of a conflict.
+
+When reading the code, there is a method ``create_snapshot_from_capability`` which downloads a capability-string from Tahoe and returns a ``RemoteSnapshot`` instance.
+
+
+General Operation
+-----------------
+
+The ``DownloaderService`` is responsible for deciding which Snapshots to download.
+The capability strings of Snapshots to download are given to ``RemoteSnapshotCacheService`` which downloads a Snapshot and all its relevant parents into memory.
```

Comment:
Yes, this probably needs additional error-handling to cover a case of "can't download a parent" .. I think that's best achieved as a followup?

Reply:
I think we want to include a discussion of that case here, even if it is not yet implemented.

Reply:
Okay, I can do that .. but also I suspect the answer is very likely to be, approximately, "we should renew leases on everything".

(Although discussion here of what to do if a parent _can't_ be downloaded would be good, because it could happen for other reasons)

Reply:
I'm more interested in what to do when we can't download a parent, since our discussion this morning suggested that for v1 we won't be renewing the leases on old snapshots.

I think it is reasonable to treat what we need to for renewal out-of-scope for this PR.

Reply:
>  "we should renew leases on everything"

If that is the answer, we may need to revisit either:
- the decision to not worry about renewing leases on old snapshots
- the design of recovery for v1 that depends on having those